codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -140,6 +143,100 @@ public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newPare
                 }
             }
         });
+
+        if (listener == null) {
+            writeAllocatedBytesVisitor = new RemoteStreamVisitor() {
+                @Override
+                public boolean visit(Http2Stream stream) {
+                    state(stream).writeAllocatedBytes();
+                    return true;
+                }
+
+                @Override
+                public void beforeIteration() {
+                }
+
+                @Override
+                public Http2Stream streamMadeConnectionNotWritable() {
+                    return null;
+                }
+            };
+            incrementWindowSizeWritabilityVisitor = initialWindowSizeWritabilityVisitor =
+                    writePendingBytesWritabilityVisitor = null;
+        } else {
+            writeAllocatedBytesVisitor = new RemoteStreamVisitor() {
+                private boolean connectionOriginalWritable;
+                private Http2Stream streamMadeConnectionNotWritable;
+                @Override
+                public boolean visit(Http2Stream stream) {
+                    AbstractState state = state(stream);
+                    int written = state.writeAllocatedBytes();
+                    if (written != -1) {
+                        try {
+                            listener.streamWritten(stream, written);
+                        } catch (RuntimeException e) {
+                            logger.error("Caught RuntimeException from listener.streamWritten", e);
+                        }
+                        // If the connection window has been observed to be made <= 0 during this iteration, then all
```

Comment:
@nmittler - Yes there are 2 loops. The `streamMadeConnectionNotWritable` exists to avoid having to do a full iteration on the second pass in this case.